Fixes nil slices leading to null fields in HTTP JSON responses

slackpad opened the pull request:
These would manifest in the HTTP output as Javascript nulls instead of empty lists, so we had unintentionally changed the interface while porting to the new state store. We added code to each HTTP endpoint to convert nil slices to empty ones so they JSON-ify properly, and we added tests to catch this in the future.

This should fix #1395 and #1406.

mitchellh commented:
Not knowing Consul's internals too much: is it possible we can do a `nil` check in the HTTP layer and do the return empty slice there? Then we can avoid allocations in the state store internals and only have them on the edges for API calls.

But if these methods are only called by the API layer, I suppose it makes no difference.

slackpad commented:
@mitchellh agree this is weird to fix at the state store level but I couldn't find any clean place to put these up in the HTTP layer without doing a bunch of post-filtering up there. I looked at the JSON marshaling and didn't see a good way to hook it in there (though that would have changed the API on other ways because we do have some `null` arrays in the JSON). I think it'll be the common case that we are going to fill these slices anyway (we were already `append`-ing onto a nil which kind of sucks since we don't know the sizes beforehand) so this won't be too bad.

mitchellh commented:
No problem, just curious, I don't think allocating the one slice with zero elements will matter that much.

slackpad commented:
Thanks for eyeballing this one! I've got a bizarro unit test failure that I didn't expect though - looking into that now :-/

slackpad commented:
LOL @mitchellh your spidey sense was right with this one. The failing test was due to a slice weirdness when I converted KV to be consistent. When a unit test tried to use the same reply object multiple times we ended up getting the underlying arrays pointed incorrectly. Looking into this I realized that KV already dealt with nil slices up at the HTTP layer, so I just changed the other HTTP endpoint code to match.

mitchellh commented:
Nice, LGTM :)
